### PR TITLE
[precompile] Let the example call keep the gradients it produced

### PR DESCRIPTION
**Pull request by bobrenjc93**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.14.0) (oldest at bottom):
* #195284
* #195215
* #195070
* __->__ #195014
* #194966
* #194952
* #194951
* #194950
* #194949
* #194948
* #194677
* #194622
* #194619
* #194549
* #194542
* #194541
* #194534
* #194488
* #194479
* #194455
* #194454
* #194452
* #194450
* #194449
* #194428
* #194427
* #194426
* #194409
* #194394
* #194393
* #194390
* #194389
* #194388
* #194387
* #194386
* #194385
* #194384
* #194383
* #194320
* #194319
* #194318
* #194317
* #194316
* #194302
* #194301
* #194300
* #194286
* #194285
* #194284
* #194283
* #194282
* #194250
* #194249
* #194248
* #194158
* #192379
* #193976
* #193697
* #192374
* #192866

precompile snapshots every example model's .grad, clears it, runs the example
calls, and restores the snapshot. That is right for the documented flow -- a
warmup step and then a capture -- where not clearing would double the gradients
the capture finds, and not restoring would leave the model changed by having
been used as an example.

It is wrong for the other flow, and silently. A caller whose example call IS
their live training step gets an artifact and no gradients: the restore writes
the pre-step values back over everything the backward just computed. On a real
model that means the dense optimizer steps on nothing for one batch while
fused sparse optimizers, which apply inside the backward and never touch .grad,
have already updated -- one batch of dense/sparse desync, with the job
reporting success.

keep_example_grads=True turns the whole snapshot/clear/restore off, so
precompile does not touch .grad at all and a gradient already present
accumulates exactly as it would in eager. Default False, because the flow that
needs the restore cannot detect that it needed it either.

Not folded into training=True, though that was tempting and would have needed
no new argument. training=True says "fn runs a backward", which the
warmup-then-capture flow also says; the two are distinguishable only by whether
the caller intends to consume these particular gradients, and that is not
something the flag already means.

Threaded through precompile_capture as well. The two entry points differed here
only by accident -- a session driven with no example_inputs has nothing to
snapshot, so it preserved gradients by doing nothing -- and that is the kind of
divergence that gets discovered the expensive way.

Two smaller things while in here, both reported alongside the above:

The on-disk form with tracer='make_fx' returned an empty list AFTER fn had
already run. The return value of that form IS the example calls' results, and a
make_fx trace has none to give, so the caller is left holding "no results" for a
region that already executed and whose obvious recovery -- run it again -- is a
second backward over storages the first one freed. Refused up front instead.

_PrecompileApi's docstring, torch/compiler/__init__.py and one comment named a
capture member, and two of them a load_package; neither exists. Removed rather
than added: the manual session form is a real design question and advertising it
as though it had been settled is worse than not mentioning it.

Test Plan:

```
python test/test_precompile.py
python test/dynamo/test_package.py
python test/functorch/test_compile_to_python.py
```

Three new tests: the default restores and the flag keeps; a gradient present
before capture accumulates to the same value two eager steps produce; and the
flag is refused for tracer='make_fx' alongside the other multi-variant knobs.
The make_fx on-disk test now asserts the region did NOT run.

Authored with the assistance of an AI coding agent.

### CODE REVIEW

**Comment by bobrenjc93** [comment]
Superseded by the canonical precompile stack #195353-#195362; this change lives in #195362 (with the guard-drop policy half in #195360). Closing to consolidate review there. The branch is kept on GitHub.
